File: Release/Source/BBC.Dna/SubNotificationEmail.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using System.Xml;
using BBC.Dna.Data;

namespace BBC.Dna.Component
{
    /// <summary>
    /// Class to store the SubNotificationEmail
    /// </summary>
    public class SubNotificationEmail : DnaInputComponent
    {
        /// <summary>
        /// Default constructor for the SubNotificationEmail component
        /// </summary>
        /// <param name="context">The Context of the DnaPage the component is created in.</param>
        public SubNotificationEmail(IInputContext context)
            : base(context)
        {
        }

        ///<summary>
        /// Checks to see if this sub editor has any allocations that they have
		///	not yet been notified about, and if so builds the email to send
		///	in order to notify them.
        /// </summary>
        /// <param name="subID">user id of the sub to create the notification for</param>
        /// <param name="toSend">returned whether or not an email to send was created</param>
        /// <param name="emailAddress">the subs email address</param>
        /// <param name="emailSubject">the subject line for the email</param>
        /// <param name="emailText">the text of the email</param>
        public void CreateNotificationEmail(int subID, ref bool toSend, ref string emailAddress, ref string emailSubject, ref string emailText)
        {
            string subName = String.Empty;
            string authorName = String.Empty;
            string subject = String.Empty;
            int authorID = 0;
            int h2g2ID = 0;
            string batchDetails = String.Empty;

            toSend = false;
            RootElement.RemoveAll();

            XmlElement email = AddElementTag(RootElement, "EMAIL");
            AddAttribute(email, "TYPE", "SUB-NOTIFICATION");

            // call the SP to fetch all the allocations that the sub has not yet been notified of
            // this also updates these allocations status and the subs last notified date
            using (IDnaDataReader dataReader = InputContext.CreateDnaDataReader("FetchAndUpdateSubsUnnotifiedAllocations"))
            {
                dataReader.AddParameter("SubID", subID);
                dataReader.AddParameter("currentsiteid", InputContext.CurrentSite.SiteID);
                dataReader.Execute();

                // Check to see if we found anything
                if (dataReader.HasRows && dataReader.Read())
                {
                    // set the email address and subject
                    toSend = true;
                    emailAddress = dataReader.GetStringNullAsEmpty("SubEmail");
                    subName = dataReader.GetStringNullAsEmpty("SubName");

                    AddTextTag(email, "EMAIL-ADDRESS", emailAddress);

                    //Add the user info with groups
                    User user = new User(InputContext);
                    user.AddUserXMLBlock(dataReader, subID, email);

                    // build up a string containing the details of the subs allocated batch
                    do
                    {
                        authorName = dataReader.GetStringNullAsEmpty("AuthorName");
                        authorID = dataReader.GetInt32NullAsZero("AuthorID");
                        subject = dataReader.GetStringNullAsEmpty("Subject");
                        h2g2ID = dataReader.GetInt32NullAsZero("h2g2ID");
                        batchDetails += "A" + h2g2ID + " '" + subject + "' by " + authorName + " (U" + authorID + ")\r\n";
                    }
                    while (dataReader.Read());

                    emailSubject = String.Empty;
                    emailText = String.Empty;

                    // fetch the template for the email
                    using (IDnaDataReader dataReader2 = InputContext.CreateDnaDataReader("fetchemailtext"))
                    {
                        dataReader2.AddParameter("SiteID", InputContext.CurrentSite.SiteID);
                        dataReader2.AddParameter("emailname", "SubAllocationsEmail");
                        dataReader2.Execute();

                        if (dataReader2.HasRows && dataReader2.Read())
                        {
                            emailSubject = dataReader2.GetStringNullAsEmpty("Subject");
                            emailText = dataReader2.GetStringNullAsEmpty("text");

                            // do any appropriate substitutions
                            emailSubject = emailSubject.Replace("++**sub_name**++", subName);
                            emailText = emailText.Replace("++**sub_name**++", subName);
                            emailText = emailText.Replace("++**batch_details**++", batchDetails);
                        }
                    }
                    AddTextTag(email, "SUBJECT", emailSubject);
                    AddTextTag(email, "TEXT", emailText);
                }
            }
        }

    }
}

File: Production/Source/Ripley/SubNotificationEmail.cpp
// SubNotificationEmail.cpp: implementation of the CSubNotificationEmail class.
//
//////////////////////////////////////////////////////////////////////

/*

Copyright British Broadcasting Corporation 2001.

This code is owned by the BBC and must not be copied, 
reproduced, reconfigured, reverse engineered, modified, 
amended, or used in any other way, including without 
limitation, duplication of the live service, without 
express prior written permission of the BBC.

The code is provided "as is" and without any express or 
implied warranties of any kind including without limitation 
as to its fitness for a particular purpose, non-infringement, 
compatibility, security or accuracy. Furthermore the BBC does 
not warrant that the code is error-free or bug-free.

All information, data etc relating to the code is confidential 
information to the BBC and as such must not be disclosed to 
any other party whatsoever.

*/


#include "stdafx.h"
#include "SubNotificationEmail.h"
#include "InputContext.h"
#include "TDVString.h"
#include "TDVDateTime.h"
#include "StoredProcedure.h"
#include "TDVAssert.h"

#if defined (_ADMIN_VERSION)

//////////////////////////////////////////////////////////////////////
// Construction/Destruction
//////////////////////////////////////////////////////////////////////

CSubNotificationEmail::CSubNotificationEmail(CInputContext& inputContext) :
	CXMLObject(inputContext)
{
	// none
}

CSubNotificationEmail::~CSubNotificationEmail()
{
	// none
}

/*********************************************************************************

	bool CSubNotificationEmail::CreateNotificationEmail(int iSubID, bool* pbToSend, CTDVString* psEmailAddress, CTDVString* psEmailSubject, CTDVString* psEmailText)

	Author:		Kim Harries
	Created:	28/11/2000
	Inputs:		iSubID - user id of the sub to create the notification for
	Outputs:	pbToSend - whether or not an email to send was created
				psEmailAddress - the subs email address
				psEmailSubject - the subject line for the email
				psEmailText - the text of the email
	Returns:	true for success or false for failure
	Purpose:	Checks to see if this sub editor has any allocations that they have
				not yet been notified about, and if so builds the email to send
				in order to notify them.

*********************************************************************************/

bool CSubNotificationEmail::CreateNotificationEmail(int iSubID, bool* pbToSend, CTDVString* psEmailAddress, CTDVString* psEmailSubject, CTDVString* psEmailText)
{
	TDVASSERT(pbToSend != NULL, "NULL pbToSend in CSubNotificationEmail::CreateNotificationEmail(...)");
	TDVASSERT(psEmailAddress != NULL, "NULL psEmailAddress in CSubNotificationEmail::CreateNotificationEmail(...)");
	TDVASSERT(psEmailText != NULL, "NULL psEmailText in CSubNotificationEmail::CreateNotificationEmail(...)");

	CStoredProcedure	SP;
	CTDVString			sEmailSubject;
	CTDVString			sEmailText;
	CTDVString			sSubName;
	CTDVString			sAuthorName;
	CTDVString			sSubject;
	int							iAuthorID = 0;
	int							ih2g2ID = 0;
	bool						bSuccess = true;
	
	CTDVString			sFirstNames;
	CTDVString			sLastName;
	CTDVString			sArea;		
	CTDVString			sSiteSuffix;
	CTDVString			sTitle;
	int							iStatus=0;
	int							iTaxonomyNode=0;
	int							iJournal = 0;
	int							iActive = 0;
		
	bool bGotSP = m_InputContext.InitialiseStoredProcedureObject(&SP);
	
	// if failed to create SP or output variables not supplied then return failure
	if (!bGotSP || pbToSend == NULL || psEmailAddress == NULL || psEmailText == NULL)
	{
		return false;
	}
	
	// otherwise call the SP to fetch all the allocations that the sub has
	// not yet been notified of
	// this also updates these allocations status and the subs last notified date
	bSuccess = bSuccess && SP.FetchAndUpdateSubsUnnotifiedAllocations(iSubID);
	
	// see if we have any fields returned
	if (SP.IsEOF())
	{
		// if nothing returned then no email to send
		*pbToSend = false;
	}
	else
	{
		// otherwise need to build the email
		*pbToSend = true;
		
		// set the email address and subject
		bSuccess = bSuccess && SP.GetField("SubEmail", *psEmailAddress);
		bSuccess = bSuccess && SP.GetField("SubName", sSubName);
		
		bSuccess = bSuccess && SP.GetField("FirstNames", sFirstNames);
		bSuccess = bSuccess && SP.GetField("LastName", sLastName);
		bSuccess = bSuccess && SP.GetField("Area", sArea);
		bSuccess = bSuccess && SP.GetField("SiteSuffix", sSiteSuffix);
		bSuccess = bSuccess && SP.GetField("Title", sTitle);

		iStatus = SP.GetIntField("Status");
		iTaxonomyNode = SP.GetIntField("TaxonomyNode");
		iJournal = SP.GetIntField("Journal");
		iActive = SP.GetIntField("Active");

		EscapeAllXML(&sFirstNames);
		EscapeAllXML(&sLastName);
		EscapeAllXML(&sArea);
		EscapeAllXML(&sSiteSuffix);
		EscapeAllXML(&sTitle);

		// build up a string containing the details of the subs allocated batch
		CTDVString	sBatchDetails = "";
		while (bSuccess && !SP.IsEOF())
		{
			bSuccess = bSuccess && SP.GetField("AuthorName", sAuthorName);
			bSuccess = bSuccess && SP.GetField("Subject", sSubject);
			iAuthorID = SP.GetIntField("AuthorID");
			ih2g2ID = SP.GetIntField("h2g2ID");
			sBatchDetails << "A" << ih2g2ID << " '" << sSubject << "' by " << sAuthorName << " (U" << iAuthorID << ")\r\n";
			SP.MoveNext();
		}

		// fetch the template for the email
		bSuccess = bSuccess && SP.FetchEmailText(m_InputContext.GetSiteID(), 	"SubAllocationsEmail", &sEmailText, &sEmailSubject);
		
		// do any appropriate substitutions
		sEmailSubject.Replace("++**sub_name**++", sSubName);
		sEmailText.Replace("++**sub_name**++", sSubName);
		sEmailText.Replace("++**batch_details**++", sBatchDetails);
		
		// set the output variables if they are present
		if (psEmailSubject != NULL)
		{
			*psEmailSubject = sEmailSubject;
		}
		if (psEmailText != NULL)
		{
			*psEmailText = sEmailText;
		}
	}
	
	// now create an XML marked up representation of this email for the XML object
	CTDVString	sXML = "";
	if (*pbToSend)
	{
		CTDVString	sEmailAddress = *psEmailAddress;

		// must escape any XML in these before outputting them
		CXMLObject::EscapeXMLText(&sEmailAddress);
		CXMLObject::EscapeXMLText(&sEmailSubject);
		CXMLObject::EscapeXMLText(&sEmailText);

		// if there is an email to send then represent it in XML		
		sXML << "<EMAIL TYPE='SUB-NOTIFICATION'>";
		sXML << "<EMAIL-ADDRESS>" << sEmailAddress << "</EMAIL-ADDRESS>";

		//get the groups to which this user belongs to 		
		CTDVString sGroupXML;		
		m_InputContext.GetUserGroups(sGroupXML, iSubID);		

		sXML << "<USER>";
			sXML << "<USERID>" << iSubID << "</USERID>";
			sXML << "<USERNAME>" << sSubName << "</USERNAME>";
			sXML << "<FIRSTNAMES>" << sFirstNames << "</FIRSTNAMES>";
			sXML << "<LASTNAME>" << sLastName << "</LASTNAME>";
			sXML << "<AREA>" << sArea << "</AREA>";
			sXML << "<STATUS>" << iStatus << "</STATUS>";
			sXML << "<TAXONOMYNODE>" << iTaxonomyNode << "</TAXONOMYNODE>";
			sXML << "<JOURNAL>" << iJournal << "</JOURNAL>";
			sXML << "<ACTIVE>" << iActive << "</ACTIVE>";
			sXML << "<SITESUFFIX>" << sSiteSuffix << "</SITESUFFIX>";
			sXML << "<TITLE>" << sTitle << "</TITLE>";
			sXML << sGroupXML;
		sXML << "</USER>";

		sXML << "<SUBJECT>" << sEmailSubject << "</SUBJECT>";
		sXML << "<TEXT>" << sEmailText << "</TEXT>";
		sXML << "</EMAIL>";
	}
	else
	{
		// if no data then just have an empty tag
		sXML << "<EMAIL TYPE='SUB-NOTIFICATION'/>";
	}
	// first destroy any existing tree
	delete m_pTree;
	m_pTree = NULL;
	// then create the new internal tree
	bSuccess = bSuccess && CreateFromXMLText(sXML);
	return bSuccess;
}

#endif // _ADMIN_VERSION
